lib/Dialect/TTNN/Transforms/TTNNWorkarounds.cpp

```diff
@@ -360,6 +361,100 @@ class TTNNOperandsWorkaroundsRewriter
   }
 };
 
+//
+// Two workarounds are implemented here to avoid issues in ttnn
+//
+// 1. all_reduce ops are broken down into reduce_scatter and all_gather ops
+// because current support of all_reduce in TTNN is not stable.
+// 2. reduce_scatter op in TTNN currently does not support two dimensional
+// tensor correctly. As a temporary workaround, we insert reshape ops front
+// and back to make the tensor as four dimensional tensor.
+class TTNNAllReduceWorkarounds : public OpRewritePattern<ttnn::AllReduceOp> {
+public:
+  using OpRewritePattern<ttnn::AllReduceOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ttnn::AllReduceOp op,
+                                PatternRewriter &rewriter) const override {
+    RankedTensorType inputType =
+        mlir::cast<RankedTensorType>(op.getInput().getType());
+    llvm::SmallVector<int64_t> inputTypeShape(inputType.getShape());
+    size_t scatter_dim = op.getScatterDim();
+    int32_t scatter_num = op.getScatterNum();
+    auto device = op.getDevice();
+    Location loc = op.getLoc();
+
+    // TODO(wooseoklee): Once it supports two dimensional tensor
+    // (https://github.com/tenstorrent/tt-metal/issues/15010), we can remove
+    // this workaround solution.
+    if (inputTypeShape.size() < 4) {
```

Comment:
Regarding this, I am creating this issue:
https://github.com/tenstorrent/tt-mlir/issues/1517

We will create a workaround infra to support these examples more easily. Once we introduce the infra for that will move this part to the appropriate place. :)

Reply:
Got it. I will track this issue as well on my side.